xdsl/dialects/experimental/stencil.py

```diff
@@ -114,42 +111,112 @@ def __iter__(self) -> Iterator[int]:
         return (e.data for e in self.array.data)
 
 
+@irdl_attr_definition
+class StencilBoundsAttr(ParametrizedAttribute, TypeAttribute):
+    name = "stencil.bounds"
+    lb: ParameterDef[IndexAttr]
+    ub: ParameterDef[IndexAttr]
+
+    def _verify(self):
+        if len(self.lb) != len(self.ub):
+            raise VerifyException(
+                "Incoherent stencil bounds: lower and upper bounds must have the same dimensionality."
+            )
+        for d in self.ub - self.lb:
+            if d <= 0:
+                raise VerifyException(
+                    "Incoherent stencil bounds: upper bound must be strictly greater than lower bound."
+                )
+
+    def __init__(self, bounds: Sequence[tuple[int | IntAttr, int | IntAttr]]):
+        lb = tuple(b[0] for b in bounds)
+        ub = tuple(b[1] for b in bounds)
+        super().__init__(
+            [
+                IndexAttr.get(*lb),
+                IndexAttr.get(*ub),
+            ]
+        )
+
+
 class StencilType(
     Generic[_FieldTypeElement], ParametrizedAttribute, TypeAttribute, builtin.ShapeType
 ):
-    shape: ParameterDef[IndexAttr]
+    bounds: ParameterDef[StencilBoundsAttr | IntAttr]
     element_type: ParameterDef[_FieldTypeElement]
 
     def get_num_dims(self) -> int:
-        return len(self.shape)
+        if isinstance(self.bounds, IntAttr):
+            return self.bounds.data
+        else:
+            return len(self.bounds.ub.array.data)
 
     def get_shape(self) -> tuple[int]:
-        return tuple(self.shape)
+        if isinstance(self.bounds, IntAttr):
+            return (-1,) * self.bounds.data
+        else:
+            return tuple(self.bounds.ub - self.bounds.lb)
 
     @staticmethod
     def parse_parameters(parser: Parser) -> list[Attribute]:
+        def parse_interval() -> tuple[int, int] | int:
+            if parser.parse_optional_punctuation("?"):
+                return -1
+            parser.parse_punctuation("[")
+            l = parser.parse_integer(False)
```

Comment:
What does False mean here? Seems like a label would be useful

Reply:
It means don't allow booleans, I can use a named arg sure !